docs/state-of-rent.md

```diff
@@ -0,0 +1,86 @@
+- During rent collection, sets the rent epoch for "rent-exempt" accounts to the `RENT_EXEMPT_RENT_EPOCH` marker as well
+- This has the side effect of not needing to update accounts every epoch anymore.
+- Note that this feature behavior didn't remove accounts from account delta hash calculations because when we run rent collection for an account, even if there are no changes made, the account is still considered "written" in that block. This behavior of writing an account without any actual changes is called a "rewrite" and the `skip_rent_rewrites` feature aims to remove rewrites.
+
+#### `validatee_fee_collector_account`
+Feature address: https://explorer.solana.com/address/prpFrMtgNmzaNzkPJg9o753fVvbHKqNrNTm76foJ2wm
+Feature issue: https://github.com/solana-labs/solana/issues/33888
+Activated on mainnet in epoch 598 (~2024-04-04)
+- Prevents new rent-paying accounts from being created by collecting fees into a non-existent validator id account
+
+#### `skip_rent_rewrites`
+Feature address: https://explorer.solana.com/address/CGB2jM8pwZkeeiXQ66kBMyBR6Np61mggL7XUsmLjVcrw
+Feature issue: https://github.com/solana-labs/solana/issues/26599
+Not yet activated on mainnet
+- When running partitioned rent collection for a block, only write the accounts that were actually changed (either rent epoch or rent amount) and only include those written accounts in the account delta hash
+- Note that this feature was modified in (https://github.com/anza-xyz/agave/pull/2910) to fix an issue where we would not rewrite account when we should have because the rent epoch was updated. This bug is actually not hit normally on a cluster like mainnet where all accounts are rent exempt. You would need to make a pre-existing rent-paying account into a rent exempt account and then rent collection would fail to set the rent epoch to the marker value `RENT_EXEMPT_RENT_EPOCH`. Another case is when we create new sysvars, builtins, and precompiles because they get created with a default rent epoch of 0.
+
+#### `disable_rent_fees_collection`
+Feature address: https://explorer.solana.com/address/CJzY83ggJHqPGDq8VisV3U91jDJLuEaALZooBrXtnnLU
+Feature issue: https://github.com/solana-labs/solana/issues/33946
+Not yet activated on mainnet
+- Stops collecting rent even if there are still rent paying accounts and in `collect_rent_from_account` we no longer consider executable / incinerator to be rent exempt and therefore won't attempt to set rent epoch to MAX
+- Therefore, new builtins / precompiles will not have their rent epoch set to MAX anymore because they don't have a rent exempt balance and are created with an initial rent epoch of 0
+- Since rent is no longer collected, no rent is paid to validators
+
+Strangeness:
+- We still run partitioned rent code on each bank just for setting rent epoch to MAX
+	- This will only ever happen after creating new sysvars or when existing rent-paying accounts become rent exempt
```

Comment:
@HaoranYi @brooksprumo it's a bit surprising to me that even after activating `disable_rent_fees_collection` we will still be using the partitioned rent code to load all accounts each epoch in order to update the rent epoch for a few edge cases since basically all accounts already have a rent epoch set to MAX.

I think we should maybe have a new feature or two which:
1. sets the MAX rent epoch for new sysvars, builtins, and precompiles
2. sets the MAX rent epoch for rent-paying accounts that become rent exempt immediately
3. stops calling rent partitioned code.

Is there an extra benefit to reading all accounts once per epoch that I'm not aware of?

Reply:
Rent collection has the side effect on updating rent_epoch. 

Initially, disable_rent_fees_collection was coded to skip both rent fees and rent_epoch update. But we think there might be a potential risk that rent paying account that don't have its rent_epoch set to MAX. So we change it to keep the rent_epoch update logic. 

Now we are sure that there is no rent-paying accounts on mnb. We should skip rent_epoch update code all together, i.e. https://github.com/anza-xyz/agave/blob/master/svm/src/account_loader.rs#L109-L122

But we should not 

> stops calling rent partitioned code.

Because stop calling rent partitioned code, will skip rewrites, which is control by the "skip rewrites" feature.

Reply:
Yeah so after the "skip rewrites" feature is activated we won't need to write accounts in the rent partitioned code anymore and it can be removed with a new feature gate. Sound good @HaoranYi?

Reply:
Yes, we can do it after skip rewrites feature is enabled.